client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -316,7 +433,21 @@ public void close() throws IOException {
         client.close();
     }
 
-    private static IOException addSuppressedException(IOException suppressedException, IOException currentException) {
+    private static boolean isSuccessfulResponse(String method, int statusCode) {
+        return statusCode < 300 || (HttpHead.METHOD_NAME.equals(method) && statusCode == 404);
+    }
+
+    private static boolean mayRetry(int statusCode) {
```

Comment:
I like `mayRetry` best out of those three options. The method's job is to say where or not it'd be ok if the client retried the request. If it were named `maybeRetry` I'd expect it to attempt the retry.